scripts/nbconverted/3_add_batch_effects.py

```diff
@@ -0,0 +1,214 @@
+        simulated_data.to_csv(batch_file, sep='\t')
+        
+    else:  
+        batch_data_df = pd.DataFrame()
+        for j in range(i):
+            
+            # CLEAN UP
+            #print('batch number: {}'.format(j))
+            #if j == 0:
+                # Randomly select samples
+            #   batch_df = new_data_decoded_df.sample(n=num_samples_per_batch, frac=None, replace=False)
+            #   batch_data_df = batch_data_df.append(batch_df)    
+            #else:
+            
+            stretch_factor = np.random.uniform(0,1)
+            
+            # Randomly select samples
+            batch_df = simulated_data.sample(n=num_samples_per_batch, frac=None, replace=False)
+            batch_df.columns = batch_df.columns.astype(str)
+
+            # Add batch effect
+            subset_genes_to_change_tile = pd.DataFrame(
+                pd.np.tile(
+                    subset_genes_to_change,
+                    (num_samples_per_batch, 1)),
+                index=batch_df.index)
+
+            offset_vector = pd.DataFrame(subset_genes_to_change_tile*stretch_factor)
+            offset_vector.columns = offset_vector.columns.astype(str)
+            batch_df = batch_df + offset_vector
+
+            # if any exceed 1 then set to 1 since gene expression is normalized
+            batch_df[batch_df>=1.0] = 1.0
+
+
+            # Append batched together
+            batch_data_df = batch_data_df.append(batch_df)
+
+            # Select a new direction (i.e. a new subset of genes to change)
+            np.random.shuffle(subset_genes_to_change)
+            
+        # Save
+        batch_data_df.to_csv(batch_file, sep='\t')
+
+
+# ## Plot batch data using UMAP
+
+# In[7]:
+
+
+# Plot generated data 
+
+for i in num_batches:
+    batch_data_file = os.path.join(
+        os.path.dirname(os.getcwd()),
+        "data",
+        "batch_simulated",
+        analysis_name,
+        "Batch_"+str(i)+".txt")
+    
+    batch_data = pd.read_table(
+        batch_data_file,
+        header=0,
+        sep='\t',
+        index_col=0)
+    
+    # UMAP embedding of decoded batch data
+    batch_data_UMAPencoded = umap.UMAP().fit_transform(batch_data)
+    batch_data_UMAPencoded_df = pd.DataFrame(data=batch_data_UMAPencoded,
+                                             index=batch_data.index,
+                                             columns=['1','2'])
+    
+        
+    g = ggplot(aes(x='1',y='2'), data=batch_data_UMAPencoded_df) +                 geom_point(alpha=0.5) +                 scale_color_brewer(type='qual', palette='Set2') +                 ggtitle("{} Batches".format(i))
+    
+    print(g)
+
+
+# ## Plot batch data using PCA
+
+# In[8]:
+
+
+# Plot generated data 
+
+for i in num_batches:
+    batch_data_file = os.path.join(
+        os.path.dirname(os.getcwd()),
+        "data",
+        "batch_simulated",
+        analysis_name,
+        "Batch_"+str(i)+".txt")
+    
+    batch_data = pd.read_table(
```

Comment:
good practice but really no need to multi-line it if the line is =<72 characters long

Reply:
I realized that its a lot easier for me to read in this format as I've started using it and this way its also consistent